### DISCUSSION TITLE
Can't get interrupts to work

### DISCUSSION THREAD

**Discussion by gizmo71** [earlephilhower/arduino-pico#276] [Q&A] [answered]
Using the Arduino IDE 1.8.15, arduino-pico 1.9.3 as the board, on a Raspberry Pi Pico (bought this week!). My first sketch, which uses a toggle switch with each side connect to digital pins 14 and 15 (physically 19 and 20) and a potentiometer attached to ADC 0 (physical pin 31). The first version used simple polling to read the input values, which it writes out via the serial port, but I thought I would try and get the two digital inputs to use interrupts instead, with no success.

```
const uint POT_PIN = A0;
const uint SWITCH1_PIN = D14;
const uint SWITCH2_PIN = D15;

volatile int s1, s2, pot;

#define USE_INTERRUPTS 0

#if USE_INTERRUPTS
void switch1() {
  s1 = digitalRead(SWITCH1_PIN) == LOW;
}

void switch2() {
  s2 = digitalRead(SWITCH2_PIN) == LOW;
}
#endif

void setup() {
  pinMode(SWITCH1_PIN, INPUT_PULLUP);
  pinMode(SWITCH2_PIN, INPUT_PULLUP);
#if USE_INTERRUPTS
  switch1();
  attachInterrupt(digitalPinToInterrupt(SWITCH1_PIN), switch1, CHANGE);
  switch2();
  attachInterrupt(digitalPinToInterrupt(SWITCH2_PIN), switch2, CHANGE);
#endif
}

void loop() {
  pot = analogRead(POT_PIN);
#if !USE_INTERRUPTS
  s1 = digitalRead(SWITCH1_PIN) == LOW;
  s2 = digitalRead(SWITCH2_PIN) == LOW;
#endif
  sleep_ms(5);
}

void setup1() {
  Serial.begin(115200);
}

void loop1() {
  sleep_ms(500);
  Serial.print(pot);
  Serial.print(", ");
  Serial.print(s1);
  Serial.print("/");
  Serial.println(s2);
}
```

(There's no fundamental reason why the serial stuff is in the second core. I've tried doing everything in the first core, with no change in outcome.)

With `USE_INTERRUPTS` set to 0, it all works swimmingly. But with it set to 1, s1/s2 never get set; the potentiometer still gets read, as you would expect.

Looking at #25/#26, I would expect interrupts to work; I'm probably doing something really stupid, but can't for the life of me see what.

Any ideas?

**Comment by grandaspanna** [comment, marked as answer]
I've definitely had interrupts working, both with the Arduino attachInterrupt() and the native  gpio_set_irq_enabled_with_callback(). I settled on the latter so I could have one interrupt handler that worked out which pin triggered the interrupt.

However, that doesn't explain your issue. Can't see an obvious problem, but in my code, I simply use the GPIO number in the digitalPinToInterrupt macro. Otherwise, I'd start by putting some LED toggling in the interrupt handler, to avoid any issues with variables. That way, you can see if the interrupt is getting called at all.

Having said all of that, I tried your code on my Pico and it works (core: 1.9.1)

**Reply by gizmo71** [reply]
Thanks, good to know I haven't done something glaringly stupid.

When you say "I tried your code on my Pico and it works", was that with `USE_INTERRUPTS` defined as `1`, or only as `0` (i.e. the non-interrupt version)? (Probably a stupid question but I always ask them anyway - experience and all that!)

I added a LED toggler in the callbacks - definitely not being called at all. I'll try the Pico SDK calls later.

**Reply by grandaspanna** [reply]
It worked with both - and I checked by putting an additional flag in the interrupt handler to see if it was invoked. One thought was that the variable was getting cleared too quickly, but it doesn't appear to be the case.

Extract:
```
#define USE_INTERRUPTS 1

#if USE_INTERRUPTS
void switch1() {
  s1 = digitalRead(SWITCH1_PIN) == LOW;
  yeet = 1;
}
```

**Reply by gizmo71** [reply]
Thanks - I'm trying the GPIO version now. I don't actually need interrupts working in the immediate case but at some point I'd like to add a few rotary encoders and I can't see how I'm going to reliably read them without something better than pure polling.

**Reply by grandaspanna** [reply]
It's odd, because your code works for me, so it's possibly something in your environment. I do have a slightly older release, maybe something broke recently?

**Reply by gizmo71** [reply]
The `gpio_set_irq_enabled_with_callback` call works so there's obviously something borked somewhere in the Arduino wrapper; thanks for your help.

**Reply by grandaspanna** [reply]
Good news/bad news. I updated to 1.9.3 and it breaks. It works in 1.9.2 though

**Reply by earlephilhower** [reply]
Can you help pinpoint the change in 1.9.3 that caused the issue?

If you could go to 1.9.3 and edit platform.txt and change
````
compiler.libstdcpp="{runtime.platform.path}/lib/libstdc++.a"
````
to
````
compiler.libstdcpp="-lstdc++"
````

and re-try?  I have a suspicion that something in the ARM binary support got busted when disabling C++ exceptions, and the above change will use the standard library instead of the no-exceptions version.

**Reply by gizmo71** [reply]
> I have a suspicion that something in the ARM binary support got busted when disabling C++ exceptions, and the above change will use the standard library instead of the no-exceptions version.

Bingo! That edit got it working with 1.9.3. 🥇
